src/Moq/ExpressionExtensions.cs

```diff
@@ -201,6 +201,18 @@ void Split(Expression e, out Expression r /* remainder */, out InvocationShape p
 					case ExpressionType.Call:  // regular method call
 					{
 						var methodCallExpression = (MethodCallExpression)e;
+
+						if (methodCallExpression.Method.IsGenericMethod)
+						{
+							foreach (var typeArgument in methodCallExpression.Method.GetGenericArguments())
+							{
+								if (typeArgument.IsTypeMatcher())
```

Comment:
Since this loop is going to run for every generic method Moq encounters, it might be a good idea to cache information about whether or not some `Type` is a type matcher.